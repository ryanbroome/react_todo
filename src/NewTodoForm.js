import React, { useState } from "react";

const NewTodoForm = (props) => {
  // Starter / clear Form Values
  const INITIAL_STATE = {
    task: "",
  };

  // create state for form data values
  const [formData, setFormData] = useState(INITIAL_STATE);

  //   keep form values current in state formData by updating on every change
  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handleSubmit = (evt) => {
    evt.preventDefault();
    props.addTodo(formData);
    setFormData(INITIAL_STATE);
  };

  return (
    <form onSubmit={handleSubmit}>
      <h3>More Todo</h3>
      <label htmlFor='task'>Todo</label>
      <input
        id='task'
        type='text'
        name='task'
        placeholder='task'
        value={formData.task}
        onChange={handleChange}></input>
      <button>Add Todo</button>
    </form>
  );
};

export default NewTodoForm;
