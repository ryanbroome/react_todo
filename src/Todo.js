import React from "react";

const Todo = ({ id, task, removeTodo }) => {
  const handleClick = (evt) => {
    const { id } = evt.target.parentElement;
    removeTodo(id);
  };
  return (
    <li
      id={id}
      key={id}
      className='Todo'>
      <p>{task}</p>
      <button onClick={handleClick}>X</button>
    </li>
  );
};

export default Todo;
