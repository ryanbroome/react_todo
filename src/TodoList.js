import React, { useState } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import { v4 as uuid } from "uuid";

const TodoList = (props) => {
  // set state of todos
  const [todos, setTodos] = useState([]);

  const addTodo = (todo) => {
    todo.id = uuid();
    setTodos((todos) => [...todos, todo]);
  };

  const removeTodo = (id) => {
    const todosCopy = [...todos.filter((todo) => todo.id !== id)];
    setTodos(todosCopy);
  };

  return (
    <>
      <NewTodoForm addTodo={addTodo} />
      <ul className='TodoList'>
        {todos.map((todo) => (
          <Todo
            id={todo.id}
            key={todo.id}
            task={todo.task}
            removeTodo={removeTodo}
          />
        ))}
      </ul>
    </>
  );
};

export default TodoList;
