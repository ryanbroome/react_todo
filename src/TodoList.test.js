import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { v4 as uuid } from "uuid";
import NewTodoForm from "./NewTodoForm";
import TodoList from "./TodoList";
import Todo from "./Todo";

// TodoList smoke test
it("renders without crashing", function () {
  render(<TodoList />);
});

// TodoList snapshot test
it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

//Todo: NewTodoForm test
it("NewTodoForm test", function () {
  const { queryByText, getByLabelText } = render(<TodoList />);
  const input = getByLabelText("Todo");
  const btn = queryByText("Add Todo");

  expect(queryByText("Wash Car")).not.toBeInTheDocument();

  fireEvent.change(input, { target: { value: "Wash Car" } });
  fireEvent.click(btn);

  expect(queryByText("Wash Car")).toBeInTheDocument();
});

// Todo smoketest
it("renders Todo without crashing", function () {
  render(<Todo id={uuid()} />);
});

// Todo snapshot test
it("Todo snapshot test", function () {
  const { asFragment } = render(<Todo />);
  expect(asFragment()).toMatchSnapshot();
});
