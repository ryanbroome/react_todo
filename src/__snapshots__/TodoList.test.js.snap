// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo snapshot test 1`] = `
<DocumentFragment>
  <li
    class="Todo"
  >
    <p />
    <button>
      X
    </button>
  </li>
</DocumentFragment>
`;

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <h3>
      More Todo
    </h3>
    <label
      for="task"
    >
      Todo
    </label>
    <input
      id="task"
      name="task"
      placeholder="task"
      type="text"
      value=""
    />
    <button>
      Add Todo
    </button>
  </form>
  <ul
    class="TodoList"
  />
</DocumentFragment>
`;
